kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/ksp/TargetTypes.kt

```diff
@@ -0,0 +1,280 @@
+    constructor = constructor,
+    properties = properties,
+    typeVariables = typeVariables,
+    isDataClass = Modifier.DATA in type.modifiers,
+    visibility = resolvedVisibility
+  )
+}
+
+private fun ClassName.withTypeArguments(arguments: List<TypeName>): TypeName {
+  return if (arguments.isEmpty()) {
+    this
+  } else {
+    this.parameterizedBy(arguments)
+  }
+}
+
+@OptIn(KspExperimental::class)
+internal fun primaryConstructor(
+  resolver: Resolver,
+  targetType: KSClassDeclaration,
+  typeParameterResolver: TypeParameterResolver,
+  logger: KSPLogger
+): TargetConstructor? {
+  val primaryConstructor = targetType.primaryConstructor ?: return null
+
+  val parameters = LinkedHashMap<String, TargetParameter>()
+  for ((index, parameter) in primaryConstructor.parameters.withIndex()) {
+    val name = parameter.name!!.getShortName()
+    parameters[name] = TargetParameter(
+      name = name,
+      index = index,
+      type = parameter.type.toTypeName(typeParameterResolver),
+      hasDefault = parameter.hasDefault,
+      qualifiers = parameter.qualifiers(resolver),
+      jsonName = parameter.jsonName()
+    )
+  }
+
+  val kmConstructorSignature: String = resolver.mapToJvmSignature(primaryConstructor)
+    ?: run {
+      logger.error("No primary constructor found.", primaryConstructor)
+      return null
+    }
+  return TargetConstructor(
+    parameters,
+    primaryConstructor.getVisibility().toKModifier() ?: KModifier.PUBLIC,
+    kmConstructorSignature
+  )
+}
+
+private fun KSAnnotated?.qualifiers(resolver: Resolver): Set<AnnotationSpec> {
+  if (this == null) return setOf()
+  return annotations
+    .filter {
+      it.annotationType.resolve().declaration.isAnnotationPresent(JsonQualifier::class)
+    }
+    .mapTo(mutableSetOf()) {
+      it.toAnnotationSpec(resolver)
+    }
+}
+
+private fun KSAnnotated?.jsonName(): String? {
+  return this?.findAnnotationWithType<Json>()?.name
+}
+
+private fun declaredProperties(
+  constructor: TargetConstructor,
+  originalType: KSClassDeclaration,
+  classDecl: KSClassDeclaration,
+  resolver: Resolver,
+  typeParameterResolver: TypeParameterResolver,
+): Map<String, TargetProperty> {
+
+  val result = mutableMapOf<String, TargetProperty>()
+  for (property in classDecl.getDeclaredProperties()) {
+    val initialType = property.type.resolve()
+    val resolvedType = if (initialType.declaration is KSTypeParameter) {
+      property.asMemberOf(originalType.asType())
+    } else {
+      initialType
+    }
+    val propertySpec = property.toPropertySpec(resolver, resolvedType, typeParameterResolver)
+    val name = propertySpec.name
+    val parameter = constructor.parameters[name]
+    result[name] = TargetProperty(
+      propertySpec = propertySpec,
+      parameter = parameter,
+      visibility = property.getVisibility().toKModifier() ?: KModifier.PUBLIC,
+      jsonName = parameter?.jsonName ?: property.jsonName()
+        ?: name.escapeDollarSigns()
```

Comment:
This `escapeDollarSigns` call reads very suspicious to me . .. why doesn’t it escape when it comes from a parameter or property JSON name

Reply:
https://github.com/square/moshi/pull/1446/files

Reply:
this is leftover from kapt (which didn't escape them) and I believe is still necessary here too. I can test, basically it shows fine in the annotation string but doesn't play nice in generated code because of Kotlin templating